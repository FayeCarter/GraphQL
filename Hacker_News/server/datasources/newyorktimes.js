import { RESTDataSource } from 'apollo-datasource-rest';
require("dotenv").config();

export class NewYorkTimesAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://hacker-news.firebaseio.com/v0/";
  }

  articleReducer({ id, byline, url, published_date, title } = {}) {
    return {
      id: `nyt-${id}`,
      title,
      author: byline,
      url,
      time: published_date,
      source: "New York Times"
    }
  }

  async getAllArticles() {
    const result = await this.get(
      `https://api.nytimes.com/svc/mostpopular/v2/viewed/1.json?api-key=${process.env.NYT_TOKEN}`
    )
    return result?.results?.map(article => this.articleReducer(article));
  }
  
}

