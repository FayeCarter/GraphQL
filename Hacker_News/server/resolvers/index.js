export const resolvers = {
  Query: {
    articleBySource: (_, { id, source }, context) => {
      console.log("context", context)
    }
  }
}