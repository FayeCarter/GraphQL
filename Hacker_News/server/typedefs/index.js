import { gql } from 'apollo-server';

export const typeDefs = gql`
  type Query {
    articleBySource(id: ID!, source: String!): Article
  }

  type Article {
    id: ID!
    title: String
    author: String
    rel: String
    time: String
    source String
  }
`;