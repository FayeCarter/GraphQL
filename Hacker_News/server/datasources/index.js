import { HackerNewsAPI } from "./hackernews";
import { NewYorkTimesAPI } from "./newyorktimes";

export default {

}