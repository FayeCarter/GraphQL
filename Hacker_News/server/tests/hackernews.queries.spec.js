import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server';
import { HackerNewsAPI } from "../datasources/hackernews";
import { resolvers } from '../resolvers';
import { typeDefs } from '../typedefs';
import { getArticlePreReducerStub } from '../fixtures/hackernews';

const GET_ARTICLE_BY_ID_AND_SOURCE = gql`
  query getArticleByIdAndSource($id: ID!, $source: String!) {
    articleBySource(id: $id, source: $source) {
      id
      title
      author
      url
      time
      source
    }
  }
`;

const GET_ARTICLES_BY_SOURCE = gql`
  query getArticlesBySource($id: ID!, $source: String!) {
    articleBySource(ids: $ids, source: $source) {
      id
      title
      author
      url
      time
      source
    }
  }
`;

const GET_ALL_ARTICLES = gql`
  query getAllArticles {
    allArticles {
      id
      title
      author
      url
      time
      source
    }
  }
`;

const constructTestServer = () => {
  const hackerNewsAPI = new HackerNewsAPI();

  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      hackernews: hackerNewsAPI
    })
  });

  return { server, hackerNewsAPI };
}

describe("[Queries.HackerNewsAPI]", () => {
  it("fetches and article from the HackerNews API", async () => {
    const { server, hackerNewsAPI } = constructTestServer();

    hackerNewsAPI.get = jest.fn(() => getArticlePreReducerStub);

    const { query } = createTestClient(server);

    const result = await query({
      query: GET_ARTICLE_BY_ID_AND_SOURCE,
      variables: { id: 21168364, source: "hackernews" }
    })

    expect(result).toMatchSnapshot();
  })
})