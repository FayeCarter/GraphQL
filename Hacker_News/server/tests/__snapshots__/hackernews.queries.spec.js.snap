// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Queries.HackerNewsAPI] fetches an array of articles from the HackerNews API 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [ValidationError: Variable "$ids" is not defined by operation "getArticlesBySource".],
    [ValidationError: Variable "$id" is never used in operation "getArticlesBySource".],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`[Queries.HackerNewsAPI] fetches an article from the HackerNews API 1`] = `
Object {
  "data": Object {
    "articleBySource": Object {
      "author": "kristianp",
      "id": "hn-21168364",
      "source": "HackerNews",
      "time": "1570314101",
      "title": "How to Stress the C# Compiler",
      "url": "https://blog.hediet.de/post/how-to-stress-the-csharp-compiler",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
